keps/sig-api-machinery/20180731-crd-pruning.md

```diff
@@ -0,0 +1,610 @@
+         "abc": 44
+       },
+       "def": 45
+     }
+   }
+   ```
+   
+   is pruned to
+   
+   ```json
+   {
+      "json": {
+        "bar": {
+          "inner": 43
+        },
+        "def": 45
+      }
+   }
+   ```
+   
+9. `additionalProperties` within JSON
+
+   ```yaml
+   type: object
+   properties:
+     json:
+       type: object
+       x-kubernetes-preserve-unknown-fields: true
+       nullable: true
+       additionalProperties:
+         type: object
+   ```
+
+   The `additionalProperties` disables pruning **at its level**, `x-kubernetes-preserve-unknown-fields: true` already has the same effect. Inside of the additional properties `x-kubernetes-preserve-unknown-fields: true` keeps being effective, i.e.
+   
+   ```json
+   {
+     "foo": 42,
+     "json": {
+       "bar": {
+         "inner":43,
+         "abc": 44
+       },
+       "def": 45
+     }
+   }
+   ```
+   
+   is pruned to
+   
+   ```json
+   {
+     "json": {
+       "bar": {
+         "inner":43,
+         "abc": 44
+       },
+       "def": 45
+     }
+   }
+   ```   
+10. embedded resource
+
+   ```yaml
+   type: object
+   properties:
+     object:
+       type: object
+       nullable: true
+       x-kubernetes-embedded-resource: true
+       x-kubernetes-preserve-unknown-fields: true
+   ```
+   
+   Here, inside of `.object` nothing is pruned with the exception of unknown fields under `.object.metadata`, i.e.
+
+   ```json
+   {
+     "foo": 42,
+     "object": {
+       "bar": 43,
+       "abc": 44,
+       "metadata": {
+         "name": "example",
+         "garbage": 45
+       }
+     }
+   }
+   ```
+      
+   is pruned to
+      
+   ```json
+   {
+      "object": {
+        "bar": 43,
+        "abc": 44,
+        "metadata": {
+          "name": "example"
+        }
+      }
+   }
+   ```
+ 
+11. implicit `metav1.TypeMeta` and `metav1.ObjectMeta`
+
+   ```yaml
+   type: object
+   ```
+
+   Pruning takes place, but `apiVersion`, `kind`, `metadata` and known fields under `metadata` are preserved, i.e.
+   
+   ```json
+   {
+     "apiVersion": "example/v1",
+     "kind": "Foo",
+     "metadata": {
+       "name": "example",
+       "garbage": 43
+     },
+     "foo": 42
+   }
+   ```
+   
+   is pruned to
+   
+   ```json
+   {
+     "apiVersion": "example/v1",
+     "kind": "Foo",
+     "metadata": {
+       "name": "example"
+     }
+   }
+   ```
+
+### Opt-in and Opt-out of Pruning on CRD Level
+
+We will add a `PruneUnknownFields` flag to `CustomResourceDefinitionSpec` of `apiextensions.k8s.io/v1beta1`:
+
+```go
+type CustomResourceDefinitionSpec struct {
+  ...
+	
+  // pruneUnknownFields enables pruning of object fields which are not
+  // specified in the OpenAPI schema. apiVersion, kind, metadata and known
+  // fields inside metadata are excluded from pruning.
+  // Defaults to false. 
+  // Setting this field to true is considered an alpha API.
+  // Note: this will default to true in version v1.
+  PruneUnknownFields bool
```

Comment:
Since "known fields" are derived from properties/additionalProperties directives in the schema, would it be clearer to name it `x-kubernetes-preserve-unknown-properties`?

Also, would it be helpful to align the name and polarity of the field and the schema extension? This would let us omit the field in v1 when set to false, which is nice.

v1beta1:
```go
// defaults to true
// +optional
PreserveUnknownProperties *bool `json:"preserveUnknownProperties,omitempty"
```

v1:
```go
// defaults to false
// +optional
PreserveUnknownProperties bool `json:"preserveUnknownProperties,omitempty"
```

Reply:
not sure about field-vs-properties. @apelisse is using "fields" in his new tags as well.

Am fine with switching polarity.

Reply:
I like matching polarity and name, I don't feel strongly about properties vs fields

Reply:
I leave it at fields for the moment. We have time to find consensus before code freeze. This is not critical.